feat(datasync): add conflicts by version

ssd71 opened the pull request:
Rather hopeful about this one :crossed_fingers: 

```typescript
const { typeDefs, resolvers, contextCreator } = createDataSync(modelDefs, db, { Comment: { enabled: true, conflictResolution: serverSideConflictResolution }});
```

## TODOS

- [x] Agreement on if this is how it should be
- [ ] Delete Conflicts?
- [ ] Tests
- [ ] Docs

wtrocki commented on packages/graphback-datasync/src/DataSyncPlugin.ts:
```diff
@@ -131,7 +128,7 @@ export class DataSyncPlugin extends GraphbackPlugin {
     });
 
-    if (this.config.useVersion) {
+    if (this.config?.modelUsesVersion[model.graphqlType.name] || false) {
```
move to variable

craicoverflow replied:
```suggestion
    if (this.config?.modelUsesVersion[model.graphqlType.name]) {
```

Not sure if that is needed. Is the `|| false` in case the key does not exist in `modelUsesVersion`? In that case the condition would be `if (false)`.

wtrocki commented on packages/graphback-datasync/src/deltaHelper.ts:
```diff
@@ -0,0 +1,77 @@
+    const { _id,  ...delta } = updateDocument;
+    const version = delta[DataSyncFieldNames.version];
+    // eslint-disable-next-line @typescript-eslint/tslint/config
+    delete delta[DataSyncFieldNames.version];delete delta[DataSyncFieldNames.lastUpdatedAt];delete delta[DataSyncFieldNames.deleted]; 
```
Why deleting those fields... and why in single line :)

ssd71 replied:
Deleting so as to ensure only the fields changed by the user are kept in the diff

wtrocki commented on packages/graphback-datasync/src/deltaHelper.ts:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  // returns final resolved update document
+  public async resolveConflictFromDelta(updateDocument: any): Promise<any> {
```
Resolving conflict algorithm for dummies.
```
1. Accept client side version form client and validate it.
2. Look for the delta table entry with id and version.
3. This entry will be our `base` for conflict resolution.
4. Fetch server side entry 
5. Calculate diff for server side entry
5. Perform conflict check (check if fields that were changed are mergeable based on the diffs)
6. if mergable perform merge
7. If not mergable (the same fields changed - perform predefined conflict resolution.
8. Call out resolution algorithm (user provided function or default ones we provide)
We can use `MostRecentWins strategy` - that is why do not remove last updated :)
9.  try to save result to database - conditial check if version did not changed since then
7. Return data to user.

Another condition - if items reached their time (TTL ;) we have no way to kick offix compatible error to user.

```

All of this logic is already implemented in:
https://github.com/aerogear/offix/tree/master/packages/offix-conflicts-client

See also offix conflict server. Obviously you are free to invent your own algorithm but based on what I see I will probably prefer us to copy paste the code that was written and used in production rather than try to invent new way :)

wtrocki commented on packages/graphback-datasync/src/helpers/createDataSync.ts:
```diff
@@ -0,0 +1,27 @@
+    dataProviderCreator: createDataSyncConflictProviderCreator(db, dataSyncConfig),
+    plugins:[
+      ...graphbackConfig.plugins || [],
+      new DataSyncPlugin({ modelUsesVersion: Object.keys(dataSyncConfig).reduce((versionMap: any, modelName: string) => {
```
Separate variable?

wtrocki replied:
Actually why we do not do this inside plugin. It feels wrong to do it here.

craicoverflow replied:
+1 it should work the same if a user wants to use `buildGraphbackAPI` directly instead of `createDataSync`

wtrocki commented on packages/graphback-datasync/src/deltaHelper.ts:
```diff
@@ -0,0 +1,77 @@
+export class MongoDeltaHelper {
+  protected collection: Collection;
+  public constructor(model: ModelDefinition, db: Db) {
+    this.collection = db.collection(getDeltaTableName(getTableName(model.graphqlType)));
```
Extract variable

wtrocki commented on packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts:
```diff
@@ -0,0 +1,96 @@
+  public async update(updateDocument: any, context: GraphbackContext): Promise<Type> {
+    const { _id } = updateDocument;
+
+    const MAX_RETRIES = 3;
```
Probably not the best place to setup this value.

wtrocki commented on packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts:
```diff
@@ -0,0 +1,96 @@
+
+        
+        if (updateDocument[DataSyncFieldNames.version] !== serverStateData[DataSyncFieldNames.version]) {
+          const conflict: ConflictStateMap = {
```
This doesn't mean conflict. 
It means only that versions are different. 
We need diffs to know if conflict occured.

Server data might have different fields than client as well.

craicoverflow commented on packages/graphback-datasync/src/helpers/createDataSync.ts:
```diff
@@ -0,0 +1,27 @@
+
+type DataSyncGraphbackConfig = Partial<GraphbackConfig>
+
+export function createDataSync(model: string | GraphQLSchema, db: Db, dataSyncConfig: { [modelName: string]: DataSyncModelConflictConfig} = {}, graphbackConfig: DataSyncGraphbackConfig = {}): GraphbackAPI {
```
```suggestion
export function createDataSyncAPI(model: string | GraphQLSchema, db: Db, dataSyncConfig: { [modelName: string]: DataSyncModelConflictConfig} = {}, graphbackConfig: DataSyncGraphbackConfig = {}): GraphbackAPI {
```

machi1990 commented on packages/graphback-datasync/src/deltaHelper.ts:
```diff
@@ -0,0 +1,46 @@
+
+  public async findBaseForConflicts(updateDocument: any) {
+    if (updateDocument._id === undefined || updateDocument[DataSyncFieldNames.version] === undefined) {
+      throw new Error('Both _id and version field are needed for finding base');
```
```suggestion
      throw new Error(`Both _id and ${DataSyncFieldNames.version} field are needed for finding base`);
```

machi1990 commented on packages/graphback-datasync/src/deltaHelper.ts:
```diff
@@ -0,0 +1,46 @@
+  return `${tableName}Delta`
+}
+
+/**
+ * M
+ */
```
comments looks incomplete here.. can we extend this a little bit. 

It is useful since we use `typedoc` to generate API reference (even though this is not part of something we'll want to expose but let's get the habit running).

machi1990 commented on packages/graphback-datasync/src/helpers/createDataSync.ts:
```diff
@@ -0,0 +1,23 @@
+export function createDataSyncAPI(model: string | GraphQLSchema, db: Db, dataSyncConfigMap: { [modelName: string]: DataSyncModelConflictConfig} = {}, graphbackConfig: DataSyncGraphbackConfig = {}): GraphbackAPI {
+
+  return buildGraphbackAPI(model, {
+    serviceCreator: createDataSyncCRUDService({ pubSub: new PubSub() }),
```
Are we going to work with a in memory message queue for now? Looks limiting.

machi1990 commented on packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts:
```diff
@@ -0,0 +1,126 @@
+import { DataSyncMongoDBDataProvider } from './DatasyncMongoDBDataProvider';
+
+export const MAX_RETRIES = 3;
+
+/**
+ * Da
```
Comments needs some meat.

craicoverflow replied:
Requires API docs

machi1990 commented on packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts:
```diff
@@ -0,0 +1,126 @@
+import { MongoDeltaHelper } from '../deltaHelper';
+import { DataSyncMongoDBDataProvider } from './DatasyncMongoDBDataProvider';
+
+export const MAX_RETRIES = 3;
```
```suggestion
export const MAX_RETRIES = 3;
```

It will be useful to be able to configure this. 

@wtrocki what do you think?

It does not have to be done in this PR. We can open an issue to chat about it.

wtrocki replied:
Yep.. Not sure why my comments are gone.

machi1990 replied:
Opened https://github.com/aerogear/graphback/issues/1856 for this.

machi1990 commented on packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts:
```diff
@@ -0,0 +1,126 @@
+  }
+
+  public async create(data: any, context: GraphbackContext): Promise<Type> {
+    const contextWithoutSelectedFields: GraphbackContext = { ...context }
+    contextWithoutSelectedFields.graphback.options.selectedFields = [];
```
Why is this needed?

ssd71 replied:
@machi1990 @craicoverflow @wtrocki I am using it to get the complete document instead of just the selected fields. This is to ensure that the complete document is stored in the delta table:
https://github.com/aerogear/graphback/blob/601fe9f40cadf9ee5617073f491d05052d6ed9e5/packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts#L37-L39

Perhaps we do not need a delta table entry for the `create` operation, wdyt :thinking:

craicoverflow replied:
Yeah you are right to do this - I was thinking that all fields are fetched anyway, but only if subscriptions are enabled for that mutation type.

However, I think you can just modify the context directly, instead of cloning it to remove the selectedFields. There should be one context per request, and it has a short lifetime so modifying it directly should be fine.

craicoverflow replied:
See https://github.com/aerogear/graphback/blob/master/packages/graphback-core/src/runtime/CRUDService.ts#L43

machi1990 commented on packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts:
```diff
@@ -0,0 +1,126 @@
+    const clientDiff: any = {};
+    const serverDiff: any = {};
+
+    let isConflict = false;
```
```suggestion
    let conflictFound = false;
```

Just a nitpick, as it looks like we are checking if we have a conflict.

machi1990 commented on packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts:
```diff
@@ -0,0 +1,126 @@
+
+    let isConflict = false;
+
+    for (const key of Object.keys(clientData)) {
+      if (base[key] !== clientData[key]) {
+        if (!ignoredKeys.includes(key)) {
+          clientDiff[key] = clientData[key];
+        }
+      }
+    }
+
+    // calculate server diff
+    for (const key of Object.keys(clientData)) {
+      if (base[key] !== serverData[key]) {
+        if (!ignoredKeys.includes(key)) {
+          serverDiff[key] = serverData[key];
+          if (clientDiff[key]) {
+            isConflict = true;
+          }
+        }
+      }
+    }
```
can this be done in one loop?

craicoverflow commented on packages/graphback-datasync/src/DataSyncPlugin.ts:
```diff
@@ -131,7 +131,8 @@ export class DataSyncPlugin extends GraphbackPlugin {
     });
 
-    if (this.config.useVersion) {
+    const modelUsesVersion = !!this.config.modelConfigMap[model.graphqlType.name]?.enabled;
```
I don't think a double-negation is required here.

```suggestion
    const modelUsesVersion = !this.config.modelConfigMap[model.graphqlType.name]?.enabled;
```

ssd71 replied:
Hmm wouldn't that make the expression true when the rest of the statement is undefined?

craicoverflow replied:
My bad, what I mean to say I don't think there is a need for **any** negation here.

craicoverflow commented on packages/graphback-datasync/src/DataSyncPlugin.ts:
```diff
@@ -144,7 +145,8 @@ export class DataSyncPlugin extends GraphbackPlugin {
     const updateInputType = schemaComposer.getITC(getInputTypeName(model.graphqlType.name, GraphbackOperationType.UPDATE));
-    if (this.config.useVersion && updateInputType) {
+    const modelUsesVersion = this.config.modelConfigMap[model.graphqlType.name]?.enabled;
+    if ((modelUsesVersion) && updateInputType) {
```
```suggestion
    if (modelUsesVersion && updateInputType) {
```

craicoverflow commented on packages/graphback-datasync/src/deltaHelper.ts:
```diff
@@ -0,0 +1,46 @@
+}
+
+/**
+ * M
```
Missing/incomplete comments

craicoverflow commented on packages/graphback-datasync/src/deltaHelper.ts:
```diff
@@ -0,0 +1,46 @@
+/**
+ * M
+ */
+export class MongoDeltaHelper {
```
> a helper class is used to assist in providing some functionality, which isn't the main goal of the application or class in which it is used. 

I think the name `MongoDeltaHelper` is too generic/non-specific. It is in fact a very integral class providing a very specific purpose. I see a helper class as being a collection of methods which do not necessarily work together or have common purpose.

craicoverflow commented on packages/graphback-datasync/src/deltaHelper.ts:
```diff
@@ -0,0 +1,46 @@
+  }
+
+  public async insertDiff(updatedDocument: any) {
+    const version = updatedDocument[DataSyncFieldNames.version];
+    const { _id } = updatedDocument;
```
I believe this could be a one-liner:

```suggestion
const { _id, [DataSyncFieldNames.version]: version } = updatedDocument;
```

craicoverflow commented on packages/graphback-datasync/src/deltaHelper.ts:
```diff
@@ -0,0 +1,46 @@
+  }
+
+  public async findBaseForConflicts(updateDocument: any) {
+    if (updateDocument._id === undefined || updateDocument[DataSyncFieldNames.version] === undefined) {
```
Should have same effect:

```suggestion
    if (!updateDocument._id || !updateDocument[DataSyncFieldNames.version]) {
```

craicoverflow commented on packages/graphback-datasync/src/deltaHelper.ts:
```diff
@@ -0,0 +1,46 @@
+    this.collectionName = getDeltaTableName(getTableName(model.graphqlType));
+  }
+
+  public async insertDiff(updatedDocument: any) {
```
We could bring better type support to this class with generics:

```ts
export class MongoDeltaHelper<T = any>
...
public async insertDiff(updatedDocument: T) {
```

craicoverflow commented on packages/graphback-datasync/src/helpers/createDataSync.ts:
```diff
@@ -0,0 +1,23 @@
+
+type DataSyncGraphbackConfig = Partial<GraphbackConfig>
+
+export function createDataSyncAPI(model: string | GraphQLSchema, db: Db, dataSyncConfigMap: { [modelName: string]: DataSyncModelConflictConfig} = {}, graphbackConfig: DataSyncGraphbackConfig = {}): GraphbackAPI {
```
I think 5 arguments is too complicated for the user API, what do you think about reducing it to max 2? 

`DataSyncGraphbackConfig` can copy the required config props (crud, plugins) from [`GraphbackAPIConfig`](https://github.com/aerogear/graphback/blob/ce7236aa84cf053d0e9090115ab5b2b95d3276e4/packages/graphback/src/buildGraphbackAPI.ts#L7). This means very minimal repetition, but greatly simplifies user experience, which is paramount IMO.

```ts
export function createDataSyncAPI(model: string | GraphQLSchema, config: DataSyncGraphbackConfig): GraphbackAPI {
```

Thoughts? Do you see any complications that might arise from this? cc @machi1990

craicoverflow commented on packages/graphback-datasync/src/helpers/createDataSync.ts:
```diff
@@ -0,0 +1,23 @@
+import { createDataSyncCRUDService } from '../services';
+import { DataSyncPlugin } from '../DataSyncPlugin';
+
+type DataSyncGraphbackConfig = Partial<GraphbackConfig>
```
`GraphbackConfig` is not used in runtime API (it is used for the `generate` command. I think you are looking for `GraphbackAPIConfig`?

craicoverflow commented on packages/graphback-datasync/src/helpers/createDataSync.ts:
```diff
@@ -0,0 +1,23 @@
+export function createDataSyncAPI(model: string | GraphQLSchema, db: Db, dataSyncConfigMap: { [modelName: string]: DataSyncModelConflictConfig} = {}, graphbackConfig: DataSyncGraphbackConfig = {}): GraphbackAPI {
+
+  return buildGraphbackAPI(model, {
+    serviceCreator: createDataSyncCRUDService({ pubSub: new PubSub() }),
```
What if user wants to use their own PubSub library?

craicoverflow replied:
`serviceCreator` should still be an (optional) option in the API config, as the user may want to customise/extend it.

craicoverflow commented on packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts:
```diff
@@ -0,0 +1,126 @@
+
+  }
+
+  public async create(data: any, context: GraphbackContext): Promise<Type> {
```
```suggestion
  public async create(data: Partial<Type>, context: GraphbackContext): Promise<Type> {
```

wtrocki replied:
```suggestion
  public async create(data: Omit<Type, '_id'>, context: GraphbackContext): Promise<Type> {
```

craicoverflow replied:
Or `Partial<Omit<Type, '_id'>>`?

wtrocki replied:
Now sure. Do you mean Create will not push every field ?

craicoverflow replied:
Actually I was thinking about optional fields but this is already handled since the fields are optional it is still technically a complete Type.

craicoverflow replied:
//ignoreme

craicoverflow commented on packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts:
```diff
@@ -0,0 +1,126 @@
+    return result;
+  }
+
+  public async update(updateDocument: any, context: GraphbackContext): Promise<Type> {
```
```suggestion
  public async update(updateDocument: Partial<Type>, context: GraphbackContext): Promise<Type> {
```

wtrocki replied:
Should be done

craicoverflow commented on packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts:
```diff
@@ -0,0 +1,126 @@
+
+  public async create(data: any, context: GraphbackContext): Promise<Type> {
+    const contextWithoutSelectedFields: GraphbackContext = { ...context }
+    contextWithoutSelectedFields.graphback.options.selectedFields = [];
```
What is this for?

wtrocki replied:
What is the purpose of this?

craicoverflow commented on packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts:
```diff
@@ -0,0 +1,126 @@
+      
+      const serverData = await this.db.collection(this.collectionName).findOne({ _id, [DataSyncFieldNames.deleted]: false});
+      const base = await this.deltaHelper.findBaseForConflicts(updateDocument);
+      const clientData = updateDocument;
```
Why use this separate variable which references `updateDocument`? I don't see `updateDocument` being modified after this (unless I missed it)

craicoverflow commented on packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts:
```diff
@@ -0,0 +1,126 @@
+
+      let resolvedUpdate = Object.assign({}, updateDocument);
+
+      const updateFilter: any = { _id, [DataSyncFieldNames.version]: serverData[DataSyncFieldNames.version] }
```
Explicit type declarations are great, but when it is `any` there is no need to do it.

```suggestion
      const updateFilter = { _id, [DataSyncFieldNames.version]: serverData[DataSyncFieldNames.version] }
```

craicoverflow commented on packages/graphback-datasync/src/providers/createDataSyncMongoDbProvider.ts:
```diff
@@ -18,3 +19,18 @@ export function createDataSyncMongoDbProvider(db: Db): (...args: any[]) => Graph
   }
 }
+
+export function createDataSyncConflictProviderCreator(db: Db, datasyncConfigMap: { [modelName: string]: DataSyncModelConflictConfig } = {}) : (...args: any[]) => GraphbackDataProvider {
```
Needs API documentation with JSDoc

wtrocki commented on packages/graphback-datasync/src/deltaHelper.ts:
```diff
@@ -0,0 +1,46 @@
+    this.collectionName = getDeltaTableName(getTableName(model.graphqlType));
+  }
+
+  public async insertDiff(updatedDocument: any) {
```
Is it diff or snapshot?

ssd71 replied:
it is snapshot

wtrocki commented on packages/graphback-datasync/src/deltaHelper.ts:
```diff
@@ -0,0 +1,46 @@
+    this.collectionName = getDeltaTableName(getTableName(model.graphqlType));
+  }
+
+  public async insertDiff(updatedDocument: any) {
```
How this will work if I:

Update just single field (userName) in the user object that has 10 fields?

ssd71 replied:
I am using the updated document to add the diff:
https://github.com/aerogear/graphback/blob/601fe9f40cadf9ee5617073f491d05052d6ed9e5/packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts#L74-L76
Should I change that?

wtrocki replied:
No. just  my bad. All good here

wtrocki commented:
This looks like way to go but it seems incomplete to me. We need to have this runnable with the example app itself.

ssd71 commented:
Thanks for the detailed review guys! :D

ssd71 commented:
> This looks like way to go but it seems incomplete to me. We need to have this runnable with the example app itself.

@wtrocki By example app do you mean the template or  datasync-starter

wtrocki commented:
Yes

ssd71 commented:
Quick question: Should I do tests/docs in this PR as well or another one?

craicoverflow commented:
Can this be taken out of draft?

craicoverflow commented:
> Quick question: Should I do tests/docs in this PR as well or another one?

Tests - yes
Docs - preferable but not mandatory since it depends on when we release.

wtrocki commented:
Couple questions:
1. Do sample app (in templates) work?
2. What will be the default conflict resolution mechanism and would it be possible to use it for customization
3. Do we support deletes?

Ping me if answer on those is yes.

ssd71 commented:
@wtrocki 

1. Template works and is configured to use Server-side conflict resolution
2. Default conflict resolution strategy when conflict is enabled is to throw an offix compatible conflict error when a conflict is detected, it is not available for customisation but custom conflict resolution can be implemented by using the `ConflictResolutionStrategy` interface
3. No, as it wasn't discussed, should I open issue about it?

wtrocki commented:
> Template works and is configured to use Server-side conflict resolution
> Default conflict resolution strategy when conflict is enabled is to throw an offix compatible conflict error when a conflict is detected, it is not available for customisation but custom conflict resolution can be implemented by using the ConflictResolutionStrategy interface

This makes no sense to me - we have enabled server side conflict resolution but we are throwing error to user to do client side resolution. 

> No, as it wasn't discussed, should I open issue about it?

Do you have any questions? We kinda need this in offix.

wtrocki commented on packages/graphback-datasync/src/util.ts:
```diff
@@ -39,3 +44,35 @@ export const DataSyncFieldNames = {
+  resolveUpdate(conflict: ConflictMetadata): any
+}
+
+export const ServerSideConflictResolution: ConflictResolutionStrategy = {
```
I would recomend to provide names for strategies. ServerFirst, LastUpdatedFirst etc.

wtrocki commented on packages/graphback-datasync/src/util.ts:
```diff
@@ -39,3 +44,35 @@ export const DataSyncFieldNames = {
+
+    const resolved: any = {};
+
+    Object.keys(clientDiff).forEach((fieldName: string) => {
```
I cannot confirm if this algorithm works. Would be possible to explain motivation here.

Conflict detected by version doesn't mean actual conflict. Our implementation should know if we have conflict or mergability. Interface should clearly represent that.

ssd71 replied:
> I cannot confirm if this algorithm works. Would be possible to explain motivation here.

This is merging the fields that weren't changed

> Conflict detected by version doesn't mean actual conflict. Our implementation should know if we have conflict or mergability. Interface should clearly represent that.

I though this was doing it:https://github.com/aerogear/graphback/blob/94c0db84ebb75ff2bce18350da8680bd144fb27b/packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts#L84-L121

wtrocki commented on packages/graphback-datasync/src/util.ts:
```diff
@@ -39,3 +44,35 @@ export const DataSyncFieldNames = {
+}
+
+export const ServerSideConflictResolution: ConflictResolutionStrategy = {
+  resolveUpdate(conflict: ConflictMetadata): any {
```
resolve would be more appropriate. Resolution for deletion and update doesn't matter at all - we simply getting different diffs and.. in deletion conflicts is obviously guaranteed but that is not something we need to be bothered.

wtrocki commented on packages/graphback-datasync/src/util.ts:
```diff
@@ -39,3 +44,35 @@ export const DataSyncFieldNames = {
+    const resolved: any = {};
+
+    Object.keys(clientDiff).forEach((fieldName: string) => {
+      if (!serverDiff[fieldName]) {
```
We really need to have tests for this part. It is clearly from running this code that with this strategy we going to lose some data. 

Maybe you wanted to use full server snapshot and then apply client side diff on top of it? 
This would be something like client side first (without checking if conflict actually occurred. 

Second approach will be to merge automatically what is mergable and pass to conflictResolutionStrategy only fields that are conflicted - I have done that before but do not like as it limits flexibility of having full view of the doc.

ssd71 replied:
> Maybe you wanted to use full server snapshot and then apply client side diff on top of it?

To me:
serverDiff: What fields have changed on the server since client passed version _version
clientDiff: What fields the client wants to update
resolved: The fields that are in clientDiff but not in serverDiff

But I see that that was wrong so I will change it

wtrocki replied:
> resolved: The fields that are in clientDiff but not in serverDiff

Ok.. So we have problem with the interface. Maybe quick chat about this. 
I thought resolved is actual final object that will be saved. I see it now when debugging. 

Generally let's have quick chat about this approach - I never seen it but this actually might make a lot of sense.

wtrocki commented on packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts:
```diff
@@ -0,0 +1,122 @@
+    if (!this.conflictConfig.conflictResolution?.resolveUpdate) {
+      this.conflictConfig.conflictResolution = {
+        resolveUpdate(conflict: ConflictMetadata) {
+          throw new ConflictError(conflict)
```
Let's disscuss. We talking here about our users holding GB'S of data in diff table with still client side resolution by default. 

IMHO server side should be default, but would love to get your opinion as well

ssd71 replied:
> IMHO server side should be default, but would love to get your opinion as well

Onboard with this

wtrocki commented:
I think to summarize what I want on this PR is:

1. Use the default server-side strategy. (Most recent wins or client wins depending on capabilities)
2. and add unit test for:

- Delete: Client side changed single field, server side deleted everything
- Merge: Client side changed single field, server side changed other field. 
- Conflict: Client and server changed the same field 

3. Implement another strategy (give people choice - like serverWins and export it.

ssd71 commented:
Added default strategy: clientWins:
https://github.com/aerogear/graphback/blob/582df32e1e784a0f999ed9df3763ba58eb56e139/packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts#L20-L22

Added some tests as best as I could manage, please confirm if these are valid or if something needs to change:
https://github.com/aerogear/graphback/blob/582df32e1e784a0f999ed9df3763ba58eb56e139/packages/graphback-datasync/tests/DataSyncConflictProviderTest.ts#L83
https://github.com/aerogear/graphback/blob/582df32e1e784a0f999ed9df3763ba58eb56e139/packages/graphback-datasync/tests/DataSyncConflictProviderTest.ts#L97
https://github.com/aerogear/graphback/blob/582df32e1e784a0f999ed9df3763ba58eb56e139/packages/graphback-datasync/tests/DataSyncConflictProviderTest.ts#L112
https://github.com/aerogear/graphback/blob/582df32e1e784a0f999ed9df3763ba58eb56e139/packages/graphback-datasync/tests/DataSyncConflictProviderTest.ts#L128
https://github.com/aerogear/graphback/blob/582df32e1e784a0f999ed9df3763ba58eb56e139/packages/graphback-datasync/tests/DataSyncConflictProviderTest.ts#L145

There is one more strategy: serverWins:
https://github.com/aerogear/graphback/blob/582df32e1e784a0f999ed9df3763ba58eb56e139/packages/graphback-datasync/src/util.ts#L64
Will write tests for it now

ssd71 commented:
Added tests for serverWins strategy:
https://github.com/aerogear/graphback/blob/ece5c2436e0ee3601143284107802a810ea6fd6b/packages/graphback-datasync/tests/DataSyncConflictProviderTest.ts#L177
https://github.com/aerogear/graphback/blob/ece5c2436e0ee3601143284107802a810ea6fd6b/packages/graphback-datasync/tests/DataSyncConflictProviderTest.ts#L191
https://github.com/aerogear/graphback/blob/ece5c2436e0ee3601143284107802a810ea6fd6b/packages/graphback-datasync/tests/DataSyncConflictProviderTest.ts#L207
https://github.com/aerogear/graphback/blob/ece5c2436e0ee3601143284107802a810ea6fd6b/packages/graphback-datasync/tests/DataSyncConflictProviderTest.ts#L220
https://github.com/aerogear/graphback/blob/ece5c2436e0ee3601143284107802a810ea6fd6b/packages/graphback-datasync/tests/DataSyncConflictProviderTest.ts#L237

ssd71 commented:
Starting to add integration test now.

Quick question: Should I change existing DataSync integration test to use createDataSyncAPI? It would save me the trouble of having to write another integration test specifically for createDataSyncAPI but It would also mean that using buildGraphbackAPI to do datasync will be untested. wdyt

ssd71 commented:
Integration test is WIP, will be back later today to:
- add more tests

wtrocki commented on packages/graphback-datasync/src/util.ts:
```diff
@@ -39,3 +44,59 @@ export const DataSyncFieldNames = {
+  },
+  resolveDelete(conflict: ConflictMetadata): any {
+
+    throw new ConflictError(conflict);
```
We might need to know what operation thrown conflict (in case someone will start batching requests on client)

Separate issue probably?

wtrocki commented on packages/graphback-datasync/src/util.ts:
```diff
@@ -39,3 +44,59 @@ export const DataSyncFieldNames = {
+  resolveDelete(conflict: ConflictMetadata): any {
+    const { serverData, clientData } = conflict;
+
+    if (serverData[DataSyncFieldNames.deleted] === true || clientData[DataSyncFieldNames.deleted] === true) {
```
We are missing deletion integration tests.

[Question] want to delete from client (on server this changed) - client wins should force delete. With current handlers it looks like deletion support makes no sense at all (we always return error to client - in both strategies.

wtrocki commented:
This looks very very good. 
I would merge this in order to start playing with this in other project but we need more integration tests.

ssd71 commented:
Fixed clientWins strategy to force deletes as well as added some more integration tests for deletion

wtrocki commented on integration/tests/runtime-workflow-datasync-conflicts.ts:
```diff
@@ -143,6 +159,103 @@ test('Client side should win when conflict occurs', async () => {
+      }
+    }
+  }});
+  console.log(errors);
```
Probably worth to assert this to be empty. :)
